axlearn/common/gradient_accumulation.py

```
@@ -172,12 +167,54 @@ def fwd_helper(
                 otherwise None.
             """
             minibatch_size = _compute_minibatch_size(inputs["input_batch"], steps=steps)
+
+            def reshape_for_scan(x: Tensor):
+                """Helper function that adds a minibatch dimension while evenly dividing
+                batches across gradient accumulation iterations.
+
+                Input dimension is [Global logical Batch Size, Sequence], this first reshaped to
+                [Minibatch Size, Steps, Sequence],
+                then transposed to [steps, Minibatch Size, Sequence] this ensures that
+                batches picked up from the global batch in a staggered pattern.
+
+                The main benefit is that this avoids extra communication incurred in reshard
+                for every minibatch.
+
+                Args:
+                    x: Tensor to be reshaped.
+
+                Returns:
+                    The reshaped tensor.
+                """
+                if x.shape[0] % minibatch_size != 0:
+                    raise ValueError(
+                        f"minibatch_size {minibatch_size} does not evenly divide "
+                        f"global batch size of {x.shape[0]}"
+                    )
+
+                x = x.reshape(minibatch_size, -1, *x.shape[1:])
+                return jnp.swapaxes(x, 0, 1)
```

I guess this retains any sharding annotations on global batch size to the mini batch size -- OOI what happens if the minibatch doesn't divide the global batch sharding? Is it silently ignored?

---

JAX will throw an exception during the tracing stage if the minibatch size does not divide the global batch. This [test](https://github.com/apple/axlearn/pull/993/files#diff-a6b44b5d010f7278c598c45f37b8a1a8bdc265eecb890207fd307e2707704d5cR116) asserts the same.